src/connector/src/sink/iceberg/prometheus/monitor_partition_writer.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use arrow_schema::SchemaRef;
+use icelake::io_v2::{
+    FanoutPartitionedWriter, FanoutPartitionedWriterBuilder, FanoutPartitionedWriterMetrics,
+    IcebergWriter, IcebergWriterBuilder,
+};
+use icelake::Result;
+use risingwave_common::metrics::LabelGuardedIntGauge;
+
+#[derive(Clone)]
+pub struct FanoutPartitionedWriterWithMetricsBuilder<B: IcebergWriterBuilder> {
+    inner: FanoutPartitionedWriterBuilder<B>,
+    partition_num: LabelGuardedIntGauge<2>,
+}
+
+impl<B: IcebergWriterBuilder> FanoutPartitionedWriterWithMetricsBuilder<B> {
+    pub fn new(
+        inner: FanoutPartitionedWriterBuilder<B>,
+        partition_num: LabelGuardedIntGauge<2>,
+    ) -> Self {
+        Self {
+            inner,
+            partition_num,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<B: IcebergWriterBuilder> IcebergWriterBuilder for FanoutPartitionedWriterWithMetricsBuilder<B>
+where
+    B::R: IcebergWriter,
+{
+    type R = FanoutPartitionedWriterWithMetrics<B>;
+
+    async fn build(self, schema: &SchemaRef) -> Result<Self::R> {
+        let writer = self.inner.build(schema).await?;
+        Ok(FanoutPartitionedWriterWithMetrics {
+            inner: writer,
+            partition_num: self.partition_num,
+            current_metrics: FanoutPartitionedWriterMetrics { partition_num: 0 },
+        })
+    }
+}
+
+pub struct FanoutPartitionedWriterWithMetrics<B: IcebergWriterBuilder>
+where
```

Comment:
Is this `where` necessary? 

I think there should already be `type R: IcebergWriter` trait bound on the `IcebergWriterBuilder` trait definition. If not, I think we should add the trait bound in the trait definition instead of here in the impl.